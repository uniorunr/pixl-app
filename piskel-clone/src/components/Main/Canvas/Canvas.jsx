import React, { Component } from 'react';
import PropTypes from 'prop-types';
import './Canvas.scss';
import moveAndPaint from './utils/paintTool';
import moveAndErase from './utils/eraserTool';
import pickTheColor from './utils/pickerTool';
import paintBucket from './utils/paintBucketTool';
import sameColor from './utils/sameColorTool';
import drawStroke from './utils/strokeTool';
import drawRectangle from './utils/rectangleTool';

const activateTool = (
  id,
  state,
  props,
  x,
  y,
  canvas,
  overlay,
  updXY,
  updInit,
) => {
  let result = null;
  switch (id) {
    case 'pen':
      result = moveAndPaint(x, y, state, props, canvas, updXY);
      break;
    case 'eraser':
      result = moveAndErase(x, y, state, props, canvas, updXY);
      break;
    case 'choose-color':
      result = pickTheColor(x, y, state, props, canvas);
      break;
    case 'paint-bucket':
      result = paintBucket(x, y, props, canvas);
      break;
    case 'paint-same-pixels':
      result = sameColor(x, y, props, canvas);
      break;
    case 'stroke':
      result = drawStroke(x, y, state, props, canvas, overlay, updInit);
      break;
    case 'rectangle':
      result = drawRectangle(x, y, state, props, canvas, overlay, updInit);
      break;
    default:
      throw new Error("tool isn't found");
  }
  return result;
};

const translate = (source, target, clear) => {
  const ctx = target.getContext('2d');
  if (!clear) {
    ctx.clearRect(0, 0, target.width, target.height);
  }
  ctx.imageSmoothingEnabled = false;
  ctx.drawImage(source, 0, 0, target.width, target.height);
};

class Canvas extends Component {
  constructor() {
    super();

    this.canvasRef = React.createRef();
    this.canvasOverlayRef = React.createRef();

    this.state = {
      cursorActive: null,
      currX: null,
      lastY: null,
      initX: null,
      initY: null,
    };
  }

  deactivateCursor = () => {
    const { cursorActive } = this.state;
    const { currToolId } = this.props;
    if (currToolId === 'stroke' || currToolId === 'rectangle') {
      const overlay = this.canvasOverlayRef.current;
      translate(this.canvasOverlayRef.current, this.canvasRef.current, true);
      const overlayCtx = overlay.getContext('2d');
      overlayCtx.clearRect(0, 0, overlay.width, overlay.height);
    }
    const frame = document.querySelector('.frame__canvas_active');
    translate(this.canvasRef.current, frame);

    if (cursorActive) {
      this.setState({
        cursorActive: false,
        initX: null,
        initY: null,
      });
    }
  };

  handleMouseDown = ({ pageX, pageY }) => {
    const { currToolId, updateColor } = this.props;
    this.setState({
      cursorActive: true,
    });
    const result = activateTool(
      currToolId,
      this.state,
      this.props,
      pageX,
      pageY,
      this.canvasRef.current,
      this.canvasOverlayRef.current,
      this.updateLastCoordinates,
      this.updateInitCoordinates,
    );
    if (result && result !== 'transparent') updateColor(result, 'primary');
  };

  handleMouseMove = ({ pageX, pageY }) => {
    const { cursorActive } = this.state;
    const { currToolId } = this.props;

    if (cursorActive) {
      activateTool(
        currToolId,
        this.state,
        this.props,
        pageX,
        pageY,
        this.canvasRef.current,
        this.canvasOverlayRef.current,
        this.updateLastCoordinates,
        this.updateInitCoordinates,
      );
      const frame = document.querySelector('.frame__canvas_active');
      translate(this.canvasRef.current, frame);
    }
  };

  updateLastCoordinates = (x, y) => {
    this.setState({
      currX: x,
      lastY: y,
    });
  };

  updateInitCoordinates = (x, y) => {
    this.setState({
      initX: x,
      initY: y,
    });
  };

  render() {
    const { width, height } = this.props;

    return (
      <section className="canvas-section">
        <canvas
          className="canvas-section__canvas"
          id="canvas"
          width={width}
          height={height}
          ref={this.canvasRef}
          onMouseDown={this.handleMouseDown}
          onMouseMove={this.handleMouseMove}
          onMouseUp={this.deactivateCursor}
          onMouseLeave={this.deactivateCursor}
        />
        <canvas
          className="canvas-section__canvas-overlay"
          id="canvas-overlay"
          width={width}
          height={height}
          ref={this.canvasOverlayRef}
          onMouseDown={this.handleMouseDown}
          onMouseMove={this.handleMouseMove}
          onMouseUp={this.deactivateCursor}
          onMouseLeave={this.deactivateCursor}
        />
      </section>
    );
  }
}

Canvas.propTypes = {
  width: PropTypes.number.isRequired,
  height: PropTypes.number.isRequired,
  currToolId: PropTypes.string.isRequired,
  updateColor: PropTypes.func.isRequired,
  // eslint-disable-next-line
  primaryColor: PropTypes.string.isRequired,
  // eslint-disable-next-line
  pixelsPerCanvas: PropTypes.number.isRequired,
};

export default Canvas;
