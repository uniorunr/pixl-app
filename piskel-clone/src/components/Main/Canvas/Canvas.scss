.canvas-section {
  align-self: flex-start;
  flex-grow: 1;
  display: flex;
  justify-content: center;

  .canvas-section__canvas {
    border: 2px solid black;
  }
}
