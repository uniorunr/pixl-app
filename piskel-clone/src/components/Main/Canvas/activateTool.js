import moveAndPaint from './utils/paintTool';
import moveAndErase from './utils/eraserTool';
import pickTheColor from './utils/pickerTool';
import paintBucket from './utils/paintBucketTool';
import sameColor from './utils/sameColorTool';
import drawStroke from './utils/strokeTool';
import drawRectangle from './utils/rectangleTool';
import drawCircle from './utils/circleTool';
import moveContext from './utils/moveTool';
import verticalMirror from './utils/verticalMirror';

const activateTool = (
  id,
  state,
  props,
  x,
  y,
  canvas,
  overlay,
  updXY,
  updInit,
  button,
) => {
  let result = null;
  switch (id) {
    case 'pen':
      result = moveAndPaint(x, y, state, props, canvas, updXY, button);
      break;
    case 'eraser':
      result = moveAndErase(x, y, state, props, canvas, updXY);
      break;
    case 'choose-color':
      result = pickTheColor(x, y, state, props, canvas);
      break;
    case 'paint-bucket':
      result = paintBucket(x, y, props, state, canvas, button);
      break;
    case 'paint-same-pixels':
      result = sameColor(x, y, props, state, canvas, button);
      break;
    case 'stroke':
      result = drawStroke(x, y, state, props, canvas, overlay, updInit, button);
      break;
    case 'rectangle':
      result = drawRectangle(x, y, state, props, canvas, overlay, updInit, button);
      break;
    case 'circle':
      result = drawCircle(x, y, state, props, canvas, overlay, updInit, button);
      break;
    case 'move':
      result = moveContext(x, y, state, props, canvas, overlay, updInit);
      break;
    case 'vertical-mirror':
      result = verticalMirror(x, y, state, props, canvas, updXY, button);
      break;
    default:
      throw new Error("tool isn't found");
  }
  return result;
};

export default activateTool;
