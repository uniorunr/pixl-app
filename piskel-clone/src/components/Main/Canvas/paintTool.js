const draw = (x, y, lastX, lastY, connectTwoPointsFunc, state, props, canvas) => {
  const { cursorActive } = state;
  const { pixelsPerCanvas, width } = props;
  const pixelSize = width / pixelsPerCanvas;
  const context = canvas.getContext('2d');
  context.strokeStyle = 'black';

  if ((Math.abs(x - lastX) > 1 || Math.abs(y - lastY) > 1) && !!cursorActive) {
    connectTwoPointsFunc(x, y, lastX, lastY, pixelSize, context);
  } else {
    const rectangle = new Path2D();
    rectangle.rect(x * pixelSize, y * pixelSize, pixelSize, pixelSize);
    context.fill(rectangle);
  }
};

const connectTwoPoints = (x, y, lastX, lastY, pixelSize, context) => {
  let currX;
  let currY;
  const dx1 = Math.abs(x - lastX);
  const dy1 = Math.abs(y - lastY);
  let px = 2 * dy1 - dx1;
  let py = 2 * dx1 - dy1;
  let xe;
  let ye;
  const rectangles = new Path2D();

  if (dy1 <= dx1) {
    if (x - lastX >= 0) {
      currX = lastX;
      currY = lastY;
      xe = x;
    } else {
      currX = x;
      currY = y;
      xe = lastX;
    }

    rectangles.rect(
      currX * pixelSize,
      currY * pixelSize,
      pixelSize,
      pixelSize,
    );

    for (let i = 0; currX < xe; i += 1) {
      currX += 1;

      if (px < 0) {
        px += 2 * dy1;
      } else {
        if (
          (x - lastX < 0 && y - lastY < 0)
          || (x - lastX > 0 && y - lastY > 0)
        ) {
          currY += 1;
        } else {
          currY -= 1;
        }
        px += 2 * (dy1 - dx1);
      }

      rectangles.rect(
        currX * pixelSize,
        currY * pixelSize,
        pixelSize,
        pixelSize,
      );
    }
  } else {
    if (y - lastY >= 0) {
      currX = lastX;
      currY = lastY;
      ye = y;
    } else {
      currX = x;
      currY = y;
      ye = lastY;
    }

    rectangles.rect(
      currX * pixelSize,
      currY * pixelSize,
      pixelSize,
      pixelSize,
    );

    for (let i = 0; currY < ye; i += 1) {
      currY += 1;
      if (py <= 0) {
        py += 2 * dx1;
      } else {
        if (
          (x - lastX < 0 && y - lastY < 0)
          || (x - lastX > 0 && y - lastY > 0)
        ) {
          currX += 1;
        } else {
          currX -= 1;
        }
        py += 2 * (dx1 - dy1);
      }

      rectangles.rect(
        currX * pixelSize,
        currY * pixelSize,
        pixelSize,
        pixelSize,
      );
    }
  }
  context.fill(rectangles);
};

const moveAndPaint = (pageX, pageY, state, props, updateCoordinates, canvas) => {
  const { currX, lastY } = state;
  const { pixelsPerCanvas, width } = props;
  const pixelSize = width / pixelsPerCanvas;

  const x = Math.floor((pageX - canvas.offsetLeft) / pixelSize);
  const y = Math.floor((pageY - canvas.offsetTop) / pixelSize);
  draw(x, y, currX, lastY, connectTwoPoints, state, props, canvas);
  updateCoordinates(x, y);
};

export default moveAndPaint;
