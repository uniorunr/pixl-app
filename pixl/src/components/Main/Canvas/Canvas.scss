.canvas-section {
  align-self: flex-start;
  flex-direction: column;
  flex-grow: 1;
  display: flex;
  justify-content: center;
  align-items: center;

  .canvas-section__wrapper {
    align-self: center;
    flex-grow: 1;
    display: flex;
    justify-content: center;
  }

  .canvas-section__canvas {
    border: 2px solid black;
    background: url("../../../assets/background.jpg");
  }

  .canvas-section__canvas-overlay {
    position: absolute;
    z-index: 1;
    border: 2px solid black;
  }
}
