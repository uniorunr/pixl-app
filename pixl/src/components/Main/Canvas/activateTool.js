import moveAndPaint from './utils/paintTool';
import moveAndErase from './utils/eraserTool';
import pickTheColor from './utils/pickerTool';
import paintBucket from './utils/paintBucketTool';
import sameColor from './utils/sameColorTool';
import drawStroke from './utils/strokeTool';
import drawRectangle from './utils/rectangleTool';
import drawCircle from './utils/circleTool';
import moveContext from './utils/moveTool';
import verticalMirror from './utils/verticalMirror';
import horizontalMirror from './utils/horizontalMirror';
import lighten from './utils/lighten';

const activateTool = (
  id,
  props,
  x,
  y,
  canvas,
  overlay,
  updXY,
  updInit,
  button,
) => {
  let result = null;
  switch (id) {
    case 'pen':
      result = moveAndPaint(x, y, props, canvas, updXY, button);
      break;
    case 'eraser':
      result = moveAndErase(x, y, props, canvas, updXY);
      break;
    case 'choose-color':
      result = pickTheColor(x, y, props, canvas, button);
      break;
    case 'paint-bucket':
      result = paintBucket(x, y, props, canvas, button);
      break;
    case 'paint-same-pixels':
      result = sameColor(x, y, props, canvas, button);
      break;
    case 'stroke':
      result = drawStroke(x, y, props, canvas, overlay, updInit, button);
      break;
    case 'rectangle':
      result = drawRectangle(x, y, props, canvas, overlay, updInit, button);
      break;
    case 'circle':
      result = drawCircle(x, y, props, canvas, overlay, updInit, button);
      break;
    case 'move':
      result = moveContext(x, y, props, canvas, overlay, updInit);
      break;
    case 'vertical-mirror':
      result = verticalMirror(x, y, props, canvas, updXY, button);
      break;
    case 'horizontal-mirror':
      result = horizontalMirror(x, y, props, canvas, updXY, button);
      break;
    case 'lighten':
      result = lighten(x, y, props, canvas, updXY, button);
      break;
    default:
      throw new Error("tool isn't found");
  }
  return result;
};

export default activateTool;
