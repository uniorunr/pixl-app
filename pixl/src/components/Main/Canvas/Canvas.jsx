import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import * as actions from '../../../actions/actions';
import CanvasInfo from './CanvasInfo/CanvasInfo';
import './Canvas.scss';
import activateTool from './activateTool';

const toolsWithOverlayUse = ['stroke', 'rectangle', 'circle', 'move'];

const translate = (source, target, clear) => {
  const ctx = target.getContext('2d');
  if (!clear) {
    ctx.clearRect(0, 0, target.width, target.height);
  }
  ctx.imageSmoothingEnabled = false;
  ctx.drawImage(source, 0, 0, target.width, target.height);
};

class Canvas extends Component {
  canvasRef = React.createRef();

  canvasOverlayRef = React.createRef();

  componentDidMount() {
    const { setCanvasRefs } = this.props;
    setCanvasRefs(this.canvasRef.current, this.canvasOverlayRef.current);
  }

  componentDidUpdate = () => {
    const { framesArray } = this.props;
    const frame = framesArray.find(element => element.classList.contains('frame__canvas_active'));
    if (frame) {
      translate(frame, this.canvasRef.current);
    }
  };

  shouldComponentUpdate = (nextProps) => {
    const {
      pixelsPerCanvas, height, width, currX, currY,
    } = this.props;

    return (
      pixelsPerCanvas !== nextProps.pixelsPerCanvas
      || height !== nextProps.height
      || width !== nextProps.width
      || currX !== nextProps.currX
      || currY !== nextProps.currY
    );
  };

  deactivateDrawing = () => {
    const {
      currToolId,
      framesData,
      framesArray,
      layerKeys,
      activeLayer,
      cursorActive,
      toggleCursorState,
      updateCanvasInitCoords,
      updateMouseButtonCode,
    } = this.props;
    if (toolsWithOverlayUse.includes(currToolId)) {
      const overlay = this.canvasOverlayRef.current;
      translate(this.canvasOverlayRef.current, this.canvasRef.current, true);
      const overlayCtx = overlay.getContext('2d');
      overlayCtx.clearRect(0, 0, overlay.width, overlay.height);
    }
    const frame = document.querySelector('.frame__canvas_active');
    translate(this.canvasRef.current, frame);

    if (cursorActive) {
      toggleCursorState(false);
      updateCanvasInitCoords(null, null);
      updateMouseButtonCode(null);
      const layerKey = `layer${layerKeys[activeLayer]}`;
      framesData[layerKey] = framesArray.map(item => item.toDataURL());
      sessionStorage.setItem('framesData', JSON.stringify(framesData));
    }
  };

  handleMouseDown = ({ pageX, pageY, button }) => {
    const {
      currToolId,
      updateColor,
      toggleCursorState,
      updateMouseButtonCode,
    } = this.props;
    toggleCursorState(true);
    updateMouseButtonCode(button);
    const result = activateTool(
      currToolId,
      this.props,
      pageX,
      pageY,
      this.canvasRef.current,
      this.canvasOverlayRef.current,
      this.updateLastCoordinates,
      this.updateInitCoordinates,
      button,
    );
    if (result && result.color !== 'transparent') {
      if (result.button === 2) {
        updateColor(result.color);
      } else {
        updateColor(result.color, 'primary');
      }
    }
  };

  handleMouseMove = ({ pageX, pageY, button }) => {
    const {
      currToolId,
      pixelsPerCanvas,
      width,
      framesArray,
      cursorActive,
      updateCanvasCurrCoords,
    } = this.props;
    const pixelSize = width / pixelsPerCanvas;
    const canvas = this.canvasRef.current;
    const x = Math.floor((pageX - canvas.offsetLeft) / pixelSize);
    const y = Math.floor((pageY - canvas.offsetTop) / pixelSize);
    updateCanvasCurrCoords(x, y);

    if (cursorActive) {
      activateTool(
        currToolId,
        this.props,
        pageX,
        pageY,
        this.canvasRef.current,
        this.canvasOverlayRef.current,
        this.updateLastCoordinates,
        this.updateInitCoordinates,
        button,
      );
      const frame = framesArray.find(element => element.classList.contains('frame__canvas_active'));
      translate(this.canvasRef.current, frame);
    }
  };

  handleRightClick = (event) => {
    event.preventDefault();
  };

  updateLastCoordinates = (x, y) => {
    const { updateCanvasCurrCoords } = this.props;
    updateCanvasCurrCoords(x, y);
  };

  updateInitCoordinates = (x, y) => {
    const { updateCanvasInitCoords } = this.props;
    updateCanvasInitCoords(x, y);
  };

  clearCanvas = () => {
    const {
      framesArray, layerKeys, activeLayer, framesData,
    } = this.props;
    const canvas = this.canvasRef.current;
    const canvasCtx = canvas.getContext('2d');
    const frame = framesArray.find(element => element.classList.contains('frame__canvas_active'));
    canvasCtx.clearRect(0, 0, canvas.width, canvas.height);
    translate(canvas, frame);

    const layerKey = `layer${layerKeys[activeLayer]}`;
    framesData[layerKey] = framesArray.map(item => item.toDataURL());
    sessionStorage.setItem('framesData', JSON.stringify(framesData));
  };

  render() {
    const { width, height } = this.props;

    return (
      <section className="canvas-section" onContextMenu={this.handleRightClick}>
        <div className="canvas-section__wrapper">
          <canvas
            className="canvas-section__canvas"
            id="canvas"
            width={width}
            height={height}
            ref={this.canvasRef}
            onMouseDown={this.handleMouseDown}
            onMouseMove={this.handleMouseMove}
            onMouseUp={this.deactivateDrawing}
            onMouseLeave={this.deactivateDrawing}
          />
          <canvas
            className="canvas-section__canvas-overlay"
            id="canvas-overlay"
            width={width}
            height={height}
            ref={this.canvasOverlayRef}
            onMouseDown={this.handleMouseDown}
            onMouseMove={this.handleMouseMove}
            onMouseUp={this.deactivateDrawing}
            onMouseLeave={this.deactivateDrawing}
          />
        </div>
        <div className="canvas-section__additional-info-wrapper">
          <button
            type="button"
            className="canvas-section__clear-button"
            onClick={this.clearCanvas}
          >
            clear
          </button>
          <CanvasInfo />
        </div>
      </section>
    );
  }
}

Canvas.propTypes = {
  width: PropTypes.number.isRequired,
  height: PropTypes.number.isRequired,
  currToolId: PropTypes.string.isRequired,
  updateColor: PropTypes.func.isRequired,
  framesArray: PropTypes.instanceOf(Array).isRequired,
  pixelsPerCanvas: PropTypes.number.isRequired,
  framesData: PropTypes.instanceOf(Object).isRequired,
  layerKeys: PropTypes.instanceOf(Array).isRequired,
  activeLayer: PropTypes.number.isRequired,
  setCanvasRefs: PropTypes.func.isRequired,
  cursorActive: PropTypes.bool,
  toggleCursorState: PropTypes.func.isRequired,
  updateCanvasInitCoords: PropTypes.func.isRequired,
  updateCanvasCurrCoords: PropTypes.func.isRequired,
  updateMouseButtonCode: PropTypes.func.isRequired,
  currX: PropTypes.number.isRequired,
  currY: PropTypes.number.isRequired,
};

Canvas.defaultProps = {
  cursorActive: null,
};

const mapStateToProps = state => ({
  currToolId: state.tools.currToolId,
  pixelsPerCanvas: state.canvas.pixelsPerCanvas,
  width: state.canvas.width,
  height: state.canvas.height,
  framesArray: state.frames.framesArray,
  primaryColor: state.colors.primaryColor,
  secondaryColor: state.colors.secondaryColor,
  framesData: state.frames.framesData,
  layerKeys: state.layers.layerKeys,
  activeLayer: state.layers.activeLayer,
  cursorActive: state.canvas.cursorActive,
  currX: state.canvas.currX,
  currY: state.canvas.currY,
  initX: state.canvas.initX,
  initY: state.canvas.initY,
  mouseButton: state.canvas.mouseButton,
});

const mapDispatchToProps = (dispatch) => {
  const {
    updateColor,
    setCanvasRefs,
    toggleCursorState,
    updateCanvasInitCoords,
    updateCanvasCurrCoords,
    updateMouseButtonCode,
  } = bindActionCreators(actions, dispatch);
  return {
    updateColor: (color, isPrimary) => {
      updateColor(color, isPrimary);
    },
    setCanvasRefs: (canvas, overlay) => {
      setCanvasRefs(canvas, overlay);
    },
    toggleCursorState: (state) => {
      toggleCursorState(state);
    },
    updateCanvasInitCoords: (x, y) => {
      updateCanvasInitCoords(x, y);
    },
    updateCanvasCurrCoords: (x, y) => {
      updateCanvasCurrCoords(x, y);
    },
    updateMouseButtonCode: (button) => {
      updateMouseButtonCode(button);
    },
  };
};

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(Canvas);
