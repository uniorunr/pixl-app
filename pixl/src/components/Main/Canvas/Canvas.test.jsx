import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { JSDOM } from 'jsdom';
import Canvas from './Canvas';
import store from '../../../stores/mainStore';

const dom = new JSDOM();
global.document = dom.window.document;
global.window = dom.window;

it('renders correctly', () => {
  const canvas = document.createElement('canvas');
  canvas.setAttribute('id', 'canvas');
  document.body.appendChild(canvas);

  const tree = renderer
    .create(
      <Provider store={store}>
        <Canvas />
      </Provider>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
