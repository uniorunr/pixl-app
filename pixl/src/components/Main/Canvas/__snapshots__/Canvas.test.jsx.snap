// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<section
  className="canvas-section"
  onContextMenu={[Function]}
>
  <div
    className="canvas-section__wrapper"
  >
    <canvas
      className="canvas-section__canvas"
      height={512}
      id="canvas"
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseMove={[Function]}
      onMouseUp={[Function]}
      width={512}
    />
    <canvas
      className="canvas-section__canvas-overlay"
      height={512}
      id="canvas-overlay"
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseMove={[Function]}
      onMouseUp={[Function]}
      width={512}
    />
  </div>
  <div
    className="canvas-section__additional-info-wrapper"
  >
    <button
      className="canvas-section__clear-button"
      onClick={[Function]}
      type="button"
    >
      clear
    </button>
    <div
      className="canvas-section__info"
    >
      <span
        className="canvas-section__size"
      >
        size: 64x64
      </span>
      <span
        className="canvas-section__position"
      >
        x: 0  y: 0
      </span>
    </div>
  </div>
</section>
`;
